### PR TITLE
[code fix] Fixed trie_tree for code quality and docs & clang-tidy error check

### PR DESCRIPTION
**Pull request by kvedala**
#### Description of Change

**Clang Tidy**
Remove `performance-unnecessary-value-param` as error but keep as warning because `clang-tidy` can auto-fix it.

**Trie Tree**
* Fixed general code quality
* ISO C++11 standard conformity
* added documentation
* fixes LGTM alert
* Closes #1057 

#### Checklist

- [x] Added description of change
- [x] Added file name matches [File name guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#New-File-Name-guidelines)
- [x] Added tests and example, test must pass
- [x] Added documentation so that the program is self-explanatory and educational - [Doxygen guidelines](https://www.doxygen.nl/manual/docblocks.html)
- [x] Relevant documentation/comments is changed or added
- [x] PR title follows semantic [commit guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#Commit-Guidelines)
- [x] Search previous suggestions before making a new one, as yours may be a duplicate.
- [x] I acknowledge that all my contributions will be made under the project's license.

Notes: 

<a href="https://gitpod.io/#https://github.com/TheAlgorithms/C-Plus-Plus/pull/1059"></a>

### CODE REVIEW

File: data_structures/trie_tree.cpp [hunk 1]
```diff
@@ -1,91 +1,171 @@
...
-            root = root->arr[j];
-        } else {
-            root->arr[j] = createNode();
-            root = root->arr[j];
+/**
+ * @brief Trie class implementation
+ * @note the function ::trie::deleteString might be erroneous
+ * @todo review the function ::trie::deleteString and the commented lines
+ */
+class trie {
```
Comment (Panquesito7) [inline data_structures/trie_tree.cpp:22]: There's no namespace.
I think it shall be added.

File: data_structures/trie_tree.cpp [hunk 2]
```diff
@@ -1,91 +1,171 @@
...
+    std::cout << "hello - " << root.search("hello", 0) << "\n";
+
+    assert(!root.search("word", 0));
+    std::cout << "word - " << root.search("word", 0) << "\n";
+
+    assert(root.search("world", 0));
+    std::cout << "world - " << root.search("world", 0) << "\n";
 }
 
 int main() {
```
Comment (Panquesito7) [inline data_structures/trie_tree.cpp:167]:
```suggestion
/**
 * Main function
 */
int main() {
```

File: data_structures/trie_tree.cpp [hunk 3]
```diff
@@ -1,91 +1,209 @@
...
-trie* createNode() {
-    trie* nn = new trie();
-    for (int i = 0; i < 26; i++) nn->arr[i] = NULL;
-    nn->isEndofWord = false;
-    return nn;
-}
+/** \namespace data_structure
+ * \brief Data-structure algorithms
+ */
+namespace data_structure {
```
Comment (Panquesito7) [inline data_structures/trie_tree.cpp:20]:
```suggestion
namespace data_structures {
```

Comment (kvedala) [reply data_structures/trie_tree.cpp:20]: in the existing code, it is marked as `data_structure`. Changing it now will create two separate namespaces.

**Pull request merged**